table/tables/partition.go

```diff
@@ -263,6 +267,49 @@ type PartitionExpr struct {
 	*ForListPruning
 }
 
+// GetPartColumnsForKeyPartition is used to get partition columns for key partition table
+func (pe *PartitionExpr) GetPartColumnsForKeyPartition(columns []*expression.Column) ([]*expression.Column, []int) {
+	schema := expression.NewSchema(columns...)
+	partCols := make([]*expression.Column, len(pe.ColumnOffset))
+	colLen := make([]int, 0, len(pe.ColumnOffset))
+	for i, offset := range pe.ColumnOffset {
+		partCols[i] = schema.Columns[offset]
+		partCols[i].Index = i
+		colLen = append(colLen, partCols[i].RetType.GetFlen())
+	}
+	return partCols, colLen
+}
+
+// LocateKeyPartitionWithSPC is used to locate the destination partition for key
+// partition table has single partition column(SPC). It's called in FastPlan process.
+func (pe *PartitionExpr) LocateKeyPartitionWithSPC(pi *model.PartitionInfo,
+	r []types.Datum) (int, error) {
+	col := &expression.Column{}
+	*col = *pe.KeyPartCols[0]
+	col.Index = 0
+	return pe.LocateKeyPartition(pi, []*expression.Column{col}, r)
+}
+
+// LocateKeyPartition is the common interface used to locate the destination partition
+func (pe *PartitionExpr) LocateKeyPartition(pi *model.PartitionInfo,
+	cols []*expression.Column, r []types.Datum) (int, error) {
+	nr1 := uint64(1)
+	nr2 := uint64(4)
+	for _, col := range cols {
+		if r[col.Index].Kind() == types.KindNull {
+			nr1 ^= (nr1 << 1) | 1
+			continue
+		}
+		data, err := r[col.Index].ToBytes()
+		if err != nil {
+			return 0, err
+		}
+		nr1, nr2 = types.CalcBytesHash(data, nr1, nr2)
+	}
+	partID := nr1 % pi.Num
+	return int(partID), nil
```

Comment:
I think any standard hash algorithm would give better distribution and probably be as fast. Here is an example of crc32:
```
	h := crc32.NewIEEE()
	for _, col := range cols {
		if r[col.Index].Kind() == types.KindNull {
			h.Write([]byte{0})
		} else {
			data, err := r[col.Index].ToBytes()
			if err != nil {
				return 0, err
			}
			h.Write(data)
		}
	}
	return int(h.Sum32() % uint32(pi.Num)), nil
```

Reply:
Yes, it's so balanced distribution for the case you proposed above.
But how can we prove that it's also suitable for other scenes. 
um...

Reply:
(Similar to my comment above). I would assume that crc32 or md5 hash sums are well tested for different datasets, probably a lot of academic papers covers it as well as other online material.
I only checked crc32 quickly and it seems like there are HW optimisations in both x86 and arm64, but I'm no expert on various hash function. But as long as it is a well known algorithm and not too bad performance, I think any matching hash algorithm is OK to use.

Reply:
I will test the performance of crc32 comparing with the alrogitm used now.
If I we choose md5, it becomes a little complicated, because the result type of md5 is `[]byte`, I need to convert it to a `int` number, it will import extra workers and face to choose the right algorithm.